Files - Stop creating an EntityFile record for custom fields

colemanw opened the pull request:
Overview
----------------------------------------
The `EntityFile` table has historically been used for 2 purposes:

1. As Attachments: many-to-many joins between files and Activities, Cases & Notes.
2. As a redundant copy of the value of custom fields of type File.

The 2nd use is confusing, inconsistent, and unnecessary. This removes it so that EntityFile records can purely be thought of as Attachments.

Before
---------
Uploading a custom file creates an EntityFile record

After
-------
Never create an EntityFile record for custom fields.
Upgrader deletes existing EntityFile records for custom fields.

Technical Details
----------------------------------------
Custom fields already store the file id as the field value, so there is no reason to also create an entityFile record.

Previously it was needed for link construction, but that was fixed via dd68ec4.

colemanw committed:
Files - Stop creating an EntityFile record for custom fields

Custom fields already store the file id as the field value,
so there is no reason to also create an entityFile record.

Previously it was needed for link construction, but that was
fixed via dd68ec4.

colemanw committed:
Files - Remove EntityFile records for custom fields

demeritcowboy commented on CRM/Core/BAO/CustomValueTable.php:
```diff
@@ -91,16 +91,6 @@ public static function create($customParams, $parentOperation = NULL) {
-              $entityFileDAO->entity_table = $field['table_name'];
-              $entityFileDAO->entity_id = $field['entity_id'];
-              $entityFileDAO->file_id = $field['file_id'];
-              $entityFileDAO->save();
```
There are still places that create links with `eid=` in them, and probably also in universe, and for any file created after this that link seems to now give a 404, because the `eid=` in the url is blank.

colemanw replied:
@demeritcowboy this PR doesn't affect the links, that was done in https://github.com/civicrm/civicrm-core/pull/32290

Can you give steps to reproduce that problem so I can look into it?

demeritcowboy replied:
1. Create a custom field of type file for cases.
2. Upload a file to that field. It's maybe slightly easier to see the problem if it's an image file.
3. After saving, note the file link gives 404, and that eid= is blank in the url.

This is working in master without the PR.

colemanw replied:
@demeritcowboy good catch. That's more related to the other PR so I made a fix for 6.1: https://github.com/civicrm/civicrm-core/pull/32326

colemanw committed:
Files - Update links to omit entity id

Followup to dd68ec4.

demeritcowboy commented:
There's now a problem that you can't delete the file from a custom field of type file, and it's related to that section of code I linked on the other PR where it keeps track of references. In addition to no longer being able to use that to keep track of those references, there's nothing in the entity_file table at all, so it doesn't even enter the next code block where it would delete.

https://github.com/civicrm/civicrm-core/blob/42340d285ae8c1c86f99b3cf4af82c747c575f8e/CRM/Core/BAO/File.php#L235-L239

For example:
1. Create a custom field of type file for cases.
2. Upload something to it.
3. Edit it and click the delete icon. It pretends like everything's fine but nothing actually happened.

demeritcowboy commented on CRM/Upgrade/Incremental/sql/6.2.alpha1.mysql.tpl:
```diff
@@ -1 +1,4 @@
 {* file to handle db changes in 6.2.alpha1 during upgrade *}
+
+{* Remove EntityFile records for custom fields; they are redundant with the value of the field *}
+DELETE FROM civicrm_entity_file WHERE entity_table LIKE "civicrm_value_%";
```
* Should probably do one of those snapshot thingies.
* On a database with a huge number of file entries this might be slow. I wonder if either it should be one of those task loops, like when the relationshipcache was first populated, or if it should skip it and tell you to do it manually if it you have a lot of entries. I don't actually know that it would be slow, just it uses `like`, and then deleting a large number of records sometimes can be.

colemanw replied:
Ok I've added the snapshot thingie.

I'm not sure why it would be slow. The LIKE operator doesn't start with a wildcard and that's usually what slows it down.

demeritcowboy replied:
Any wildcard will slow it down if it can't use an index, but it's also the deletion of the rows themselves. I'll see if I can try it out somewhere big.

demeritcowboy replied:
Sorry this PR is still on my list, just haven't been able to review it yet this week.

colemanw commented:
> @demeritcowboy There's now a problem that you can't delete the file from a custom field of type file

Ok I think that's because it's attempting to delete the entityFile rather than simply deleting the file. I think that code can be simplified now.

colemanw committed:
Files - Fix deleting files from file fields

Files in custom fields do not have a corresponding entityFile record.
Instead we null the custom field value and then delete the file if no other references.
This also fixes attachment deletion to consider other types of references
before deleting files.

colemanw commented:
@demeritcowboy let me amend that... I think the code _should_ be simplified but this PR is already getting too big so I just patched the existing wonky file delete functions to continue to work without an entityFile record.

colemanw commented:
@demeritcowboy so you're saying that for every file we upload to a custom field we should create a row in `civicrm_file` and another one in `civicrm_entity_file` and then store the id in the custom value table as well? Can you say more about why that would be needed?

demeritcowboy commented:
Mainly because it would allow consistency in any code that handles `instances` of files, instead of always having to do "if custom field, look in this other table". And the custom_value_X records are already handled by the custom field subsystem.

colemanw commented:
I guess we're looking at this differently @demeritcowboy. Here's how I see it:
- `civicrm_file` tracks the file itself
- `civicrm_entity_file` is a n-to-n bridge table specifically for *file attachments*, e.g. one activity, multiple file attachments
- custom fields do not need this as they already have a 1-1 FK to `civicrm_file`

So I was thinking of this change as improving consistency and the existing behavior as inconsistent.
Currently, `civicrm_entity_file` is used for attachments... but also non-attachments.
Currently, the `civicrm_entity_file.entity_table` is wide-open and can't enforce ACLs because it's used for more than just attachments.
Currently, custom fields have a direct FK to the file... sort of, but you also have to worry about managing this other dangling record that doesn't really do anything but if you mess with it bad things will happen.

Being able to consistently rely on `civicrm_entity_file` to contain attachments, and being able to easily work with the values of file custom fields, would unblock some stuff I'm doing with FormBuilder right now... but if it's a bad idea then I'll have to rethink it.

@jensschuppe also might have an opinion on this based on what he wrote in https://lab.civicrm.org/dev/core/-/issues/5135

demeritcowboy commented:
I'm not 100% opposed to custom fields working differently, but since my civireport analogy doesn't seem persuasive let me explore another angle. I'll avoid using the word attachment since I see now you're using that to mean specifically "instances of files that _aren't referenced by custom fields_".

In drupal(8+) and wordpress, there is a media library. You can then add media to whatever including fields of type file. The records in the media library may or may not be connected to anything. The media record is analogous to civicrm_file.

When you connect it to something, this is analogous to making a record in civicrm_entity_file.

Now, there's a valid argument civicrm_entity_file isn't currently the right structure for handling references for custom file fields. Ok.

But what if I have a file in my media library, and I connect it to an activity, and then also to a custom file field on a contact. You can't currently do this in the UI, but theoretically you could, so just pretend for the moment. And you can do something like connect the same file to two activities which is the problem that comes up here in my previous comment. In order to avoid unrecoverable data loss, you have to keep track of these references. It's certainly still technically possible to do even if you record custom field references in another table (civicrm_value_XXX), but to me, the logical way of recording references is all in the same place in the civicrm_entity_file table.

But, if you want to keep what we could call "attachment references to files in the media library" separate and unrelated to "custom field references to files in the media libary", ok, but we need to avoid data loss from code deleting things before the reference count is 0. I would say that's my main concern here.

colemanw commented:
@demeritcowboy ok, sure, I agree we should support that scenario (even if the UI doesn't, we should at least write a unit test for it).
I think my proposed schema will work though:

- 1 file on disk
    - 1 record in `civicrm_file`: `id=123`
- 2 activities
    - 2 records in `civicrm_entity_file` (1 per activity) with `file_id=123`
- 1 custom field
    - 1 record in `civicrm_value_my_custom` with `my_file_field=123`

Now observe:

1. Delete first activity
    - 1st record in `civicrm_entity_file` will be deleted
2. Delete second activity
    - 2nd record in `civicrm_entity_file` will be deleted
    - File is not deleted
3. Delete custom data
    - civicrm_file`: `id=123` will be deleted
    - file on disk will be deleted

This can all work because with every delete we'll check `CRM_Core_DAO_File::getReferenceCounts()` and only delete the file when the count is zero.

If that sounds reasonable to you, I can write a unit test for that scenario and do whatever fixup is needed to the BAO to ensure it works.

demeritcowboy commented:
Appreciate you working through this. It _can_ work, yes, but I just want to check we're thinking the same thing. When you said earlier

> custom fields [...] have a 1-1 FK to civicrm_file

It's not true that they are 1-1. They are 1-many. The same civicrm_file id can be in more than one civicrm_value_X row. It's just hard to see this currently because the records being created sometimes (before the PR) are wrong. So

* 1 custom field
  * 1 record in civicrm_value_my_custom with my_file_field=123

is partly right. But you can have two different custom_value_X.entity_id's where custom_value_X.something_file_27 have the same file_id value. I can't tell immediately just by looking at getReferenceCounts to see if it already handles that, but that might be an example of something that needs updating.

An example that you can do in the UI is the example scenario I gave earlier with an activity with a custom file field, where the with-contact has two people, and you choose to make separate activities. What civi is supposed to be doing is creating one civicrm_file record. I don't know when that broke but doesn't seem recent. (Even outside of what we're talking about, two civicrm_file records with the same uri can be problematic.) I haven't put up a fix since it depends where this goes.

colemanw committed:
Api4 - Add unit test for getRefCountTotal

jensschuppe commented:
I'll have a closer look at that with @dontub because I feel we have been relying on `EntityFile` entities also for files in custom fields in extensions and deleting those records would not be an option.

Just as a quick reaction: I agree with @demeritcowboy to see it as *file is being used by an entity* regardless of how it is connected technically (either as an attachment or in a custom field, or maybe even with a manually created record which isn't visible anywhere in the UI).

My point in [dev/core#5135](https://lab.civicrm.org/dev/core/-/issues/5135) was that records in `civicrm_entity_file` for files in custom fields have an inconsistency in that they use the `civicrm_value_*` table in the `entity_table` column while having the custom value's `entity_id` and not its `id` in the `entity_id` column of `civicrm_entity_file` - so that's not a proper *dynamic FK* and thus does not allow joins from `EntityFile` to the actual entity the file is being used by.

I'd argue that `EntityFile` entites should always link a file ID to an actual entity ID (i.e. not the *CustomValue* entity). This would keep the redundant tracking of file IDs for files in custom fields, but I look at them as different concepts:
* the `civicrm_value_*` record for the technical fact that the file is in that custom field
* the `civicrm_entity_file` record for the semantic fact that the file is being used/attached to a specific entity

dontub commented:
> I'll have a closer look at that with @dontub because I feel we have been relying on EntityFile entities also for files in custom fields in extensions and deleting those records would not be an option.

I couldn't find a spot, where this PR would lead to problems, yet...

However, I think it make sense to track which entity uses which file. If consequently done, it would allow to find orphaned files which might be deleted. (Drupal uses a similar approach.)

colemanw commented:
I'm all for tracking the use of files, but I think adding entityFile records is the wrong way to do it. I'm working on some unit tests to demonstrate that they can be tracked solely via their FKs.

colemanw committed:
Api3 - Fix file.delete to work independently of entityFile records

colemanw commented:
@demeritcowboy I've rebased this for 6.3 and fixed [the bug you identified here](https://github.com/civicrm/civicrm-core/pull/32319#issuecomment-2745774461).  In the process I found [yet another bug ](https://lab.civicrm.org/dev/core/-/issues/5827)(oh joy) but it was pre-existing & I worked around it in this PR.

So now, if you create 2 or more activities that share the same custom file, you can delete the file from each activity, and the file itself will not be deleted until you delete it from every activity.

Not that the UI supports this, but deleting attachments will also now check more thoroughly in case the attachment is also shared with a custom field value.

dontub commented:
> I'm working on some unit tests to demonstrate that they can be tracked solely via their FKs.

So the detection of orphaned files need to be done like this?

```php
$fileIds = File::get(FALSE)->addSelect('id')->execute()->column('id');
foreach ($fileIds as $fileId) {
  if (0 === CoreUtil::getRefCountTotal('File', $fileId)) {
    // Orphaned file.
  }
}
```

I'm wondering what the performance of this approach would be if there are many files. Maybe it would make sense to add the possibility to check for reference existence (instead of counting) to reduce the number of queries...

demeritcowboy commented:
There's a lot here but most of it is pre-existing. I'd say the main blocker to merging is the effect on multivalue fields. It nulls out all the other records for the same contact if you edit a record and delete the file from the edit screen.

And then not blockers to merge but the two other main things are:
* the existing bug where in a couple places it creates a duplicate civicrm_file record but with the same uri, which now can potentially be more dangerous, and
* I think we should bring back the reassign test because what it's testing is still valid, and the entity is actually a different entity so you want to check it gets copied over.

Legend: Horsies! It's the standard horsey icon rating system.

Thing | Result | Comments
-- | -- | --
Create activity for two with-contacts. Choose separate   activities option. Attachment. | 🐂 | This is difficult to   evaluate because a pre-existing bug is creating two file entries with the   same uri. It should be creating one file entry and the entity_file entries   should use that same file id for both.   https://lab.civicrm.org/dev/core/-/issues/5810
Create activity for two with-contacts. Choose separate   activities option. Custom file field. | 🐎 | ok
Create case activity with custom field field. | 🐎 | ok
File on case with custom file field. | 🐃 | It creates two file   entries with the same uri. Might be slightly different than earlier issue   because that wasn't a custom field. (Pre-existing)
Create case with custom file field. | 🦙 | ok (subject to   preexisting issue https://lab.civicrm.org/dev/core/-/issues/4749 - it works   during create, but crashes if you add the file later)
Create case with custom file field on open case (so actually an   activity custom field). | 🐎 | ok
Merge case | 🐃 🐂   🐎 | Case A has the file.   Merge case A into case B => the file is still only on case A and that case   is now deleted. Same happens if there's a custom file field on open case   activity. Pre-existing but not sure from when.   https://lab.civicrm.org/dev/core/-/issues/5835.  Scenario 2: Case B has the file. Merge case A into case B => ok.
Reassign case | 🐃🐂 | Custom field on case   is ok but like earlier issue it creates another file entry with the same uri   if the custom field is on the activity. In this case it's ok that it creates   a copy because it's a new case, but it's never ok that the uri is the same.   Pre-existing.          But I don't think we should delete the existing test completely, because   while the internal storage isn't the same anymore, the point of the test was   to make sure the file doesn't go missing on the new case, which is still a   valid thing to check.
Merge contact with custom file field | 🎠 | This is better than   pre-PR because before the value_X table retains the reference to the old   contact but the entity_file no longer has it, so it's a bit mismatched, but   after the PR the missing entity_file one doesn't matter.
Edit custom file on activity by uploading a new file | 🐂 | It works, and the old   db row is deleted, but the file is still in the filesystem.
Delete activity with a custom file. | 🐂 | Both the db row and   file in the filesystem are still there. Already a known issue   https://lab.civicrm.org/dev/core/-/issues/3448
Email with attachment to multiple people. | 🐎 | ok
Multivalue custom file fields | 🐃🐃 | **Editing the record   and deleting the file nulls out all the other records for the contact.**          Deleting a record deletes from the value_X table but leaves the file record   and the physical file. This is probably similar to the earlier issue.
Uploading same file to separate activity custom fields. | 🐎 | ok because the uri is   different.

demeritcowboy commented:
Also I didn't review api calls or what extensions might be doing/expecting. I only clicked around the UI.

demeritcowboy commented on CRM/Upgrade/Incremental/php/SixThree.php:
```diff
@@ -29,6 +29,18 @@ class CRM_Upgrade_Incremental_php_SixThree extends CRM_Upgrade_Incremental_Base
     $this->addTask(ts('Upgrade DB to %1: SQL', [1 => $rev]), 'runSql', $rev);
+
+    // Delete non-attachment rows in batches of 5000
+    $fileCount = CRM_Core_DAO::singleValueQuery('SELECT COUNT(*) FROM civicrm_entity_file WHERE entity_table LIKE "civicrm_value_%"');
```
This might make it slower now, because of both count() and it's still running the LIKE query.

I did do some testing earlier on a medium size db and actually your original where it just does the delete without needing a batch seems to be fine. Sorry.

colemanw replied:
IMO the batches will be slightly slower because of the extra count query up-front, but it's for good reason: a slightly slower upgrade is preferable to a crashing upgrade!

colemanw commented:
Thanks @demeritcowboy for the thorough review and the oh-so-easy-to-understand horsie icon rating system.
I've just pushed up a fix to the regression you identified (for multi-record custom fields).
I feel like this could be merged at this point and the additional bugs tackled in sequels, prequels, spin-offs, live-action reboots, etc.
What say you?

demeritcowboy commented:
Thanks will take a look.

demeritcowboy commented:
I agree it's mergeable. Hopefully no prequels since that would add time-travel complexity.

I'm still not sure it's the optimal direction to go, but pending any last minute thoughts from anyone it's saddle-up time 🏇

colemanw commented:
Yee-haw 🤠
